include/standalone_gutters.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+#include "work_queue.h"
+#include "types.h"
+#include "buffering_system.h"
+
+/**
+ * In-memory wrapper to offer the same interface as a buffer tree.
+ */
+class StandAloneGutters : public BufferingSystem {
+private:
+  uint32_t buffer_size; // size of a buffer (including metadata)
+  WorkQueue *wq;
+  std::vector<std::vector<node_id_t>> buffers; // array dump of numbers for performance:
+                                               // DO NOT try to access directly!
+
+  /**
+   * Flushes the corresponding buffer to the queue.
+   * @param buffer      a pointer to the head of the buffer to flush.
+   * @param num_bytes   the number of bytes to flush.
+   */
+  void flush(node_id_t* buffer, uint32_t num_bytes);
+
+  /**
+   * Use buffering.conf configuration file to determine parameters of the StandAloneGutters
+   * Sets the following variables
+   * Queue_Factor :   The number of queue slots per worker removing data from the queue
+   * Size_Factor  :   Decrease the amount of bytes used per node by this multiplicative factor
+   */
+  void configure();
+
+  // configuration variables
+  uint32_t queue_factor; // number of elements in queue is this factor * num_workers
+  float gutter_factor;   // factor which increases/decreases the leaf gutter size
+public:
+  /**
+   * Constructs a new .
+   * @param size        the total length of a buffer, in updates.
+   * @param nodes       number of nodes in the graph.
+   * @param workers     the number of workers which will be removing batches
+   */
+  StandAloneGutters(node_id_t nodes, int workers);
+
+  ~StandAloneGutters();
+
+  /**
+   * Puts an update into the data structure.
+   * @param upd the edge update.
+   * @return nothing.
+   */
+  insert_ret_t insert(const update_t &upd);
+
+  /**
+   * Ask the buffer queue for data and sleep if necessary until it is available.
+   * @param data        to store the fetched data.
+   * @return            true if got valid data, false if unable to get data.
+   */
+  bool get_data(data_ret_t& data);
```

Comment:
how does this perform in comparison to just returning an `optional`?

Reply:
No idea, though currently the GutterTree doesn't rely on boost and adding an optional would introduce a dependency.

Reply:
this is a dependency that our main system already has so it’s ok

Reply:
Yeah but avoiding multiple installations of the same dependency is for some reason non-trivial. For instance we have to install gtest twice.